appsec/src/helper/sampler.hpp

```diff
@@ -19,96 +19,35 @@ namespace dds {
 static const double min_rate = 0.0001;
 class sampler {
 public:
-    sampler(std::shared_ptr<service_config> service_config)
-        : service_config_(std::move(service_config))
+    sampler(double sampler_rate)
     {
-        set_sampler_rate(service_config_->get_request_sample_rate());
-    }
-    class scope {
-    public:
-        explicit scope(std::atomic<bool> &concurrent) : concurrent_(&concurrent)
-        {
-            concurrent_->store(true, std::memory_order_relaxed);
-        }
-
-        scope(const scope &) = delete;
-        scope &operator=(const scope &) = delete;
-        scope(scope &&oth) noexcept
-        {
-            concurrent_ = oth.concurrent_;
-            oth.concurrent_ = nullptr;
-        }
-        scope &operator=(scope &&oth)
-        {
-            concurrent_ = oth.concurrent_;
-            oth.concurrent_ = nullptr;
-
-            return *this;
-        }
-
-        ~scope()
-        {
-            if (concurrent_ != nullptr) {
-                concurrent_->store(false, std::memory_order_relaxed);
-            }
-        }
-
-    protected:
-        std::atomic<bool> *concurrent_;
-    };
-
-    std::optional<scope> get()
-    {
-        const std::lock_guard<std::mutex> lock_guard(mtx_);
-
-        std::optional<scope> result = std::nullopt;
-
-        if (sample_rate_ !=
-            valid_sample_rate(service_config_->get_request_sample_rate())) {
-            set_sampler_rate(service_config_->get_request_sample_rate());
-        }
-
-        if (!concurrent_ && floor(request_ * sample_rate_) !=
-                                floor((request_ + 1) * sample_rate_)) {
-            result = {scope{concurrent_}};
-        }
-
-        request_++;
-
-        return result;
-    }
-
-protected:
-    static double valid_sample_rate(double sampler_rate)
-    {
-        // NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
         if (sampler_rate <= 0) {
             sampler_rate = 0;
         } else if (sampler_rate > 1) {
             sampler_rate = 1;
         } else if (sampler_rate < min_rate) {
             sampler_rate = min_rate;
         }
-        // NOLINTEND(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
 
-        return sampler_rate;
+        sample_rate_ = sampler_rate;
     }
 
-    void set_sampler_rate(double sampler_rate)
+    bool get()
     {
-        sampler_rate = valid_sample_rate(sampler_rate);
-
-        if (sampler_rate == sample_rate_) {
-            return;
+        bool result = false;
+        if (floor(request_ * sample_rate_) !=
+            floor((request_ + 1) * sample_rate_)) {
+            result = true;
         }
 
-        request_ = 0;
-        sample_rate_ = sampler_rate;
+        request_.fetch_add(1, std::memory_order_relaxed);
```

Comment:
you're doing three fetches of request_; each can yield different values. Instead, you should use the result of `fetch_add`